ddtrace/ext/ci.py

```diff
@@ -41,28 +43,35 @@
 
 
 def _normalize_ref(name):
+    # type: (Optional[str]) -> Optional[str]
     return _RE_TAGS.sub("", _RE_ORIGIN.sub("", _RE_REFS.sub("", name))) if name is not None else None
 
 
 def _filter_sensitive_info(url):
+    # type: (Optional[str]) -> Optional[str]
     return _RE_URL.sub("\\1", url) if url is not None else None
 
 
 def tags(env=None):
+    # type: (Optional[MutableMapping[str, str]]) -> MutableMapping[str, str]
+    """Extract and set tags from provider environ."""
     env = os.environ if env is None else env
-    tags = {}
+    tags = {}  # type: MutableMapping[str, Optional[str]]
     for key, extract in PROVIDERS:
         if key in env:
             tags = extract(env)
             break
 
+    git_info = git.extract_git_metadata()
+    # Tags collected from CI providers takes precedence over extracted git metadata
```

Comment:
is it weird to mix provider and non-provider values for a set of tags?

e.g. should we do either add all the git tags from `extract_git_metadata` or all of the git tags from the provider, rather than mix/match?

I think if I understand this correctly, if a provider has some of the git tags, but not all, then we can update from `extract_git_metadata`?

Reply:
Some (not all) CI providers provide git information that can include (but not always) git commit author name/email, commit sha, commit message, and git branch/tag. It really depends on the provider 😦 

These are always supposed to have a higher priority than the information we can extract from the git executable, and so in this step we extract the git info and then update the CI provided tags with any new info we found from the git executables. It definitely is messy, but from what I've seen from other tracers I can't think of a cleaner way to handle this.

Reply:
makes sense, I just wanted to be sure this was expected.

that is provider info takes precedence, and we add any missing provider git tags from git executable.